dynosaur_derive/src/traits.rs

```diff
@@ -57,3 +60,109 @@ pub(crate) fn struct_trait_params(item_trait: &ItemTrait) -> StructTraitParams {
         },
     }
 }
+
+pub(crate) fn self_receiver(item_trait: &ItemTrait) -> SelfReceiver {
+    let mut visitor = ReceiverVisitor(SelfReceiver {
+        shared_ref: false,
+        mut_ref: false,
+        owned: false,
+        box_ref: false,
+        other: false,
+    });
+    visitor.visit_item_trait(item_trait);
+    visitor.0
+}
+
+#[derive(PartialEq, Eq)]
+pub(crate) struct SelfReceiver {
+    pub(crate) shared_ref: bool,
+    pub(crate) mut_ref: bool,
+    pub(crate) owned: bool,
+    pub(crate) box_ref: bool,
+    pub(crate) other: bool,
+}
+
+impl SelfReceiver {
+    pub(crate) fn should_gen_ref(&self) -> bool {
+        matches!(
+            *self,
+            SelfReceiver {
+                shared_ref: _,
+                mut_ref: false,
+                owned: false,
+                box_ref: false,
+                other: false,
+            }
+        )
+    }
+
+    pub(crate) fn should_gen_mut_ref(&self) -> bool {
+        matches!(
+            *self,
+            SelfReceiver {
+                shared_ref: true,
+                mut_ref: _,
+                owned: false,
+                box_ref: false,
+                other: false,
+            } | SelfReceiver {
+                shared_ref: _,
+                mut_ref: true,
+                owned: false,
+                box_ref: false,
+                other: false,
+            } | SelfReceiver {
+                shared_ref: false,
+                mut_ref: false,
+                owned: false,
+                box_ref: false,
+                other: false,
+            }
+        )
+    }
+
+    pub(crate) fn should_gen_box_self(&self) -> bool {
+        matches!(
+            *self,
+            SelfReceiver {
+                shared_ref: _,
+                mut_ref: _,
+                owned: _,
+                box_ref: _,
+                other: false,
+            }
+        )
+    }
+}
+
+struct ReceiverVisitor(SelfReceiver);
+
+impl Visit<'_> for ReceiverVisitor {
```

Comment:
This can be a TODO for later but I think we need to be more picky. This doesn't handle `self: &&Self` or `self: &mut Box<Self>` for example.